### ISSUE TITLE
Undocumented build issues on macOS

### ISSUE THREAD

**Issue by LandGrey** [CodeIntelligenceTesting/jazzer#116]
same to [issue-21](https://github.com/CodeIntelligenceTesting/jazzer/issues/21) 

MacOS version
```
10.15.7
```

java -version
```
java version "1.8.0_241"
Java(TM) SE Runtime Environment (build 1.8.0_241-b07)
Java HotSpot(TM) 64-Bit Server VM (build 25.241-b07, mixed mode)
```

bazel --version
```
2.1.1-homebrew
```

clang --version
```
Apple clang version 11.0.3 (clang-1103.0.32.29)
Target: x86_64-apple-darwin19.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

run :
```
git clone https://github.com/CodeIntelligenceTesting/jazzer
cd jazzer
bazel run //:jazzer -- --cp=target.jar
```

 error:
```
INFO: Options provided by the client:
  Inherited 'common' options: --isatty=1 --terminal_columns=149
INFO: Reading rc options for 'run' from /Users/xxx/jazzer/.bazelrc:
  Inherited 'build' options: --incompatible_strict_action_env -c opt --cxxopt=-std=c++17 --action_env=CC=clang --java_language_version=8 --tool_java_language_version=9
ERROR: Unrecognized option: --java_language_version=8
```

**Comment by fmeum** [comment]
Thanks for the detailed report!

It looks like your version of Bazel is behind the most recent one by two major versions. Could you try using Bazelisk https://github.com/bazelbuild/bazelisk/releases/tag/v1.9.0 to automatically install the required release 4.1.0 or just install it manually?

**Comment by LandGrey** [comment]
> Thanks for the detailed report!
> 
> It looks like your version of Bazel is behind the most recent one by two major versions. Could you try using Bazelisk https://github.com/bazelbuild/bazelisk/releases/tag/v1.9.0 to automatically install the required release 4.1.0 or just install it manually?

yeah, I downloaded https://releases.bazel.build/4.1.0/release/bazel-4.1.0-darwin-x86_64 and 
run again :
```
./bazel-4.1.0-darwin-x86_64 run //:jazzer -- --cp=target.jar
```
there maybe error again [issue-64](https://github.com/CodeIntelligenceTesting/jazzer/issues/64):

```
ERROR: /Users/xxxx/jazzer/third_party/jni/BUILD.bazel:25:10: no such target '@local_jdk//:lib/server/libjvm.dylib': target 'lib/server/libjvm.dylib' not declared in package '' (did you mean 'jre/lib/server/libjvm.dylib'?) defined by /private/var/tmp/_bazel_xxxx/11c10d0a646678bf7a01653298e24bce/external/local_jdk/BUILD.bazel and referenced by '//third_party/jni:libjvm_osx'
ERROR: Analysis of target '//:jazzer' failed; build aborted: Analysis failed
INFO: Elapsed time: 0.070s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (1 packages loaded, 0 targets configured)
FAILED: Build did NOT complete successfully (1 packages loaded, 0 targets configured)
    Fetching @kt_java_stub_template; fetching
    Fetching @com_github_jetbrains_kotlin; fetching
```

**Comment by fmeum** [comment]
Could you try adding the flag `--//third_party/jni:jdk_8` after `run`?

Alternatively, you can use a newer JDK (e.g. 15).

**Comment by LandGrey** [comment]
> Could you try adding the flag `--//third_party/jni:jdk_8` after `run`?
> 
> Alternatively, you can use a newer JDK (e.g. 15).

I try this:
```
./bazel-4.1.0-darwin-x86_64 run --sandbox_debug --verbose_failures --//third_party/jni:jdk_8 //:jazzer -- --cp=target.jar
```
there output others error:
```
INFO: Analyzed target //:jazzer (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
ERROR: /Users/xxxx/jazzer/driver/BUILD.bazel:4:11: Compiling driver/sanitizer_hooks_with_pc.cpp failed: (Aborted): sandbox-exec failed: error executing command
  (cd /private/var/tmp/_bazel_xxxx/11c10d0a646678bf7a01653298e24bce/sandbox/darwin-sandbox/39/execroot/jazzer && \
  exec env - \
    APPLE_SDK_PLATFORM=MacOSX \
    APPLE_SDK_VERSION_OVERRIDE=10.15 \
    CC=clang \
    DEVELOPER_DIR=/Applications/Xcode.app/Contents/Developer \
    PATH=/bin:/usr/bin:/usr/local/bin \
    SDKROOT=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk \
    TMPDIR=/var/folders/8k/ww7dx7mn52j68s39p9msk_hc0000gn/T/ \
    XCODE_VERSION_OVERRIDE=11.3.0.11C29 \
  /usr/bin/sandbox-exec -f /private/var/tmp/_bazel_xxxx/11c10d0a646678bf7a01653298e24bce/sandbox/darwin-sandbox/39/sandbox.sb /var/tmp/_bazel_xxxx/install/d07238c5957d0addf241b6be07f3c14b/process-wrapper '--timeout=0' '--kill_delay=15' external/local_config_cc/wrapped_clang '-D_FORTIFY_SOURCE=1' -fstack-protector -fcolor-diagnostics -Wall -Wthread-safety -Wself-assign -fno-omit-frame-pointer -g0 -O2 -DNDEBUG '-DNS_BLOCK_ASSERTIONS=1' '-std=c++11' 'DEBUG_PREFIX_MAP_PWD=.' -iquote . -iquote bazel-out/darwin-opt-ST-a0b8f5bc1936/bin -MD -MF bazel-out/darwin-opt-ST-a0b8f5bc1936/bin/driver/_objs/sanitizer_hooks_with_pc/sanitizer_hooks_with_pc.d '-frandom-seed=bazel-out/darwin-opt-ST-a0b8f5bc1936/bin/driver/_objs/sanitizer_hooks_with_pc/sanitizer_hooks_with_pc.o' -isysroot __BAZEL_XCODE_SDKROOT__ -F__BAZEL_XCODE_SDKROOT__/System/Library/Frameworks -F__BAZEL_XCODE_DEVELOPER_DIR__/Platforms/MacOSX.platform/Developer/Library/Frameworks '-mmacosx-version-min=10.15' '-std=c++17' '-std=c++17' -no-canonical-prefixes -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -target x86_64-apple-macosx -c driver/sanitizer_hooks_with_pc.cpp -o bazel-out/darwin-opt-ST-a0b8f5bc1936/bin/driver/_objs/sanitizer_hooks_with_pc/sanitizer_hooks_with_pc.o)
driver/sanitizer_hooks_with_pc.cpp:51:16: error: 'asm goto' constructs are not supported yet
  asm volatile goto(
               ^
1 error generated.
Error in child process '/usr/bin/xcrun'. 1
Target //:jazzer failed to build
INFO: Elapsed time: 0.344s, Critical Path: 0.20s
INFO: 19 processes: 19 internal.
FAILED: Build did NOT complete successfully
FAILED: Build did NOT complete successfully
```

**Comment by fmeum** [comment]
Which version of clang and XCode are you using? `asm goto` requires at least clang 9.

**Comment by LandGrey** [comment]
> Which version of clang are you using? `asm goto` requires at least clang 9.

clang --version

```
Apple clang version 11.0.3 (clang-1103.0.32.29)
Target: x86_64-apple-darwin19.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

**Comment by fmeum** [comment]
`asm goto` is supported as of clang 9, so that should work. Maybe Bazel detects an older version of Xcode/clang and uses that? You could set `CC` to the full path of that clang 11 before invoking Bazel and see whether that helps.

Sorry that this is so difficult for you. I don't have access to a Mac myself and everything runs fine in the GitHub Actions pipeline on macOS 10.15, so I can't debug the issues you are facing myself.

**Comment by LandGrey** [comment]
> `asm goto` is supported as of clang 9, so that should work. Maybe Bazel detects an older version of Xcode/clang and uses that? You could set `CC` to the full path of that clang 11 before invoking Bazel and see whether that helps.
> 
> Sorry that this is so difficult for you. I don't have access to a Mac myself and everything runs fine in the GitHub Actions pipeline on macOS 10.15, so I can't debug the issues you are facing myself.

Thank you for your patience. I would try again.

**Comment by LandGrey** [comment]
I modify `.bazelrc` file, change 
```
build --action_env=CC=clang
```
to clang absolute path
```
build --action_env=CC=/Library/Developer/CommandLineTools/usr/bin/clang
```
There are same errors output.

**Comment by fmeum** [comment]
I see that you are using Xcode 11. I don't know why, but that may be the issue according to https://stackoverflow.com/a/64341856. Can you update to Xcode 12 and ensure that its command-line tools are used as described in the comments to that answer?

**Comment by LandGrey** [comment]
> I see that you are using Xcode 11. I don't know why, but that may be the issue according to https://stackoverflow.com/a/64341856. Can you update to Xcode 12 and ensure that its command-line tools are used as described in the comments to that answer?

I will try maybe later.

**Comment by LandGrey** [comment]
I reinstall xcode 12, and there's no `error: 'asm goto' constructs are not supported yet` error. 

but still output:
```
INFO: Analyzed target //:jazzer (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //:jazzer up-to-date:
  bazel-bin/jazzer
INFO: Elapsed time: 0.092s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
INFO: Running command line: bazel-bin/jazzer driver/jazzer_driver '--cp=target-INFO: Build completed successfully, 1 total action
mkdir: /tmp/jazzer: File exists
Could not find jazzer_agent_deploy.jar. Please provide the pathname via the --agent_path flag.
```
then I add `--agent_path` flag
```
./bazel-4.1.0-darwin-x86_64 run --sandbox_debug --verbose_failures --//third_party/jni:jdk_8 //:jazzer -- --cp=target.jar --agent_path=/xxx/jazzer_release/jazzer_api_deploy.jar
```

still remain:
```
INFO: Analyzed target //:jazzer (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //:jazzer up-to-date:
  bazel-bin/jazzer
INFO: Elapsed time: 0.083s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
INFO: Running command line: bazel-bin/jazzer driver/jazzer_driver '--cp=target-INFO: Build completed successfully, 1 total action
mkdir: /tmp/jazzer: File exists
Failed to find Premain-Class manifest attribute in /xxxx/jazzer_release/jazzer_api_deploy.jar
Error occurred during initialization of VM
agent library failed to init: instrument
```

**Comment by fmeum** [comment]
The `--agent_path` needs to point to the `jazzer_agent_deploy.jar`, not the `jazzer_api_deploy.jar`. Does that work for you?

`//:jazzer` should set the `--agent_path` for you, so this failure is somewhat unexpected. Once we get it to work, I will look into this.

**Comment by LandGrey** [comment]
> The `--agent_path` needs to point to the `jazzer_agent_deploy.jar`, not the `jazzer_api_deploy.jar`. Does that work for you?
> 
> `//:jazzer` should set the `--agent_path` for you, so this failure is somewhat unexpected. Once we get it to work, I will look into this.

sorry, I change to `jazzer_agent_deploy.jar`，error output:
```
Exception in thread "main" java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:386)
	at sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:401)
Caused by: java.lang.NoSuchMethodError: java.nio.file.Path.of(Ljava/lang/String;[Ljava/lang/String;)Ljava/nio/file/Path;
	at com.code_intelligence.jazzer.agent.ClassNameGlobber.includes(RuntimeInstrumentor.kt:60)
	at com.code_intelligence.jazzer.agent.Agent.premain(Agent.kt:92)
	... 6 more
FATAL ERROR in native method: processing of -javaagent failed
/private/var/tmp/_bazel_xxxx/11c10d0a646678bf7a01653298e24bce/execroot/jazzer/bazel-out/darwin-opt/bin/jazzer: line 18: 52790 Abort trap: 6           driver/jazzer_driver -artifact_prefix=/tmp/jazzer/ --reproducer_path=/tmp/jazzer --cp=target.jar --agent_path=/xxxx/jazzer_agent_deploy.jar
```

**Comment by fmeum** [comment]
This has been fixed in https://github.com/CodeIntelligenceTesting/jazzer/commit/2b93d5799f2a676c12212a93cdf545077ac99142#diff-44220d2c982a1ee0621b41e8eb042cda4f6b94f794a343a177fc531be4d4f944. Try `git pull` and run the command again with latest `master`.

**Comment by LandGrey** [comment]
> This has been fixed in [2b93d57#diff-44220d2c982a1ee0621b41e8eb042cda4f6b94f794a343a177fc531be4d4f944](https://github.com/CodeIntelligenceTesting/jazzer/commit/2b93d5799f2a676c12212a93cdf545077ac99142#diff-44220d2c982a1ee0621b41e8eb042cda4f6b94f794a343a177fc531be4d4f944). Try `git pull` and run the command again with latest `master`.

I am use `--agent_path` flag specify releases [v0.9.1](https://github.com/CodeIntelligenceTesting/jazzer/releases/tag/v0.9.1) jazzer_agent_deploy.jar. 
If I not use `--agent_path` flag, there are output tips
```
Could not find jazzer_agent_deploy.jar. Please provide the pathname via the --agent_path flag.
```
so, I am use latest code, but there are not found latest `jazzer_agent_deploy.jar`

**Comment by fmeum** [comment]
Sorry, I meant that the Java exception you are seeing would be solved by that commit, not the missing agent.

You can (and should) always build Jazzer from its most recent commit as that's what is running in OSS-Fuzz. Our releases always lag behind quite a bit unfortunately.

**Comment by LandGrey** [comment]
> Sorry, I meant that the Java exception you are seeing would be solved by that commit, not the missing agent.
> 
> You can (and should) always build Jazzer from its most recent commit as that's what is running in OSS-Fuzz. Our releases always lag behind quite a bit unfortunately.

How can I manually by myself build `jazzer_agent_deploy.jar`，I do not kown about bazel ~

**Comment by fmeum** [comment]
@LandGrey Glad to hear that! The issue that the path to `jazzer_agent_deploy.jar` had to be passed in manually was a recent regression that has now been fixed in #118.
